odl/space/base_tensors.py

```diff
@@ -354,6 +355,77 @@ def __contains__(self, other):
         """
         return getattr(other, 'space', None) == self
 
+    def __getitem__(self, indices):
+        """Return ``self[indices]``.
+
+        For all supported cases, indexing is implemented such that for an
+        element ``x in space``, the statement ::
+
+            x[indices] in space[indices]
+
+        is ``True``.
+
+        Space indexing works with
+
+        - integers,
+        - `slice` objects,
+        - index arrays, i.e., 1D array-like objects containing integers,
+
+        and combinations of the above. It does not work with boolean arrays
+        or more advanced "fancy indexing".
+
+        .. note::
+            This method is a default implementation that propagates only
+            ``shape`` and ``dtype``. Subclasses with more properties
+            need to override the method.
+
+        Examples
+        --------
+        A single integer slices along the first axis (index does not
+        matter as long as it lies within the bounds):
+
+        >>> rn = odl.rn((3, 4, 5, 6))
+        >>> rn[0]
+        rn((4, 5, 6))
+        >>> rn[2]
+        rn((4, 5, 6))
+
+        Multiple indices slice into corresponding axes from the left:
+
+        >>> rn[0, 0]
+        rn((5, 6))
+        >>> rn[0, 1, 1]
+        rn(6)
+
+        Ellipsis (``...``) and ``slice(None)`` (``:``) can be used to keep
+        one or several axes intact:
+
+        >>> rn[0, :, 1, :]
+        rn((4, 6))
+        >>> rn[..., 0, 0]
+        rn((3, 4))
+
+        With slices, parts of axes can be selected:
+
+        >>> rn[0, :3, 1:4, ::2]
+        rn((3, 3, 3))
+
+        Array-like objects (must all have the same 1D shape) of integers are
+        treated as follows: if their common length is ``n``, a new axis
+        of length ``n`` is created at the position of the leftmost index
+        array, and all index array axes are collapsed (Note: this is
+        not so useful for spaces, more so for elements):
+
+        >>> rn[0, 0, [0, 1, 0, 2], :]
+        rn((4, 6))
+        >>> rn[:, [1, 1], [0, 2], :]
+        rn((3, 2, 6))
+        >>> rn[[2, 0], [3, 3], [0, 1], [5, 2]]
+        rn(2)
+        """
+        new_shape, _, _ = simulate_slicing(self.shape, indices)
+        return type(self)(shape=new_shape, dtype=self.dtype)
```

Comment:
Not quite sure we want a default implementation here, this could be very implementation specific (see e.g. weighting)

Reply:
Fair enough, I'll remove the default implementation.